core/domain/state_domain.py

```diff
@@ -916,8 +940,1085 @@ def get_all_outcomes(self) -> List[Outcome]:
+                        rule_spec.inputs['f'])
+                if rule_spec.rule_type == 'IsEqualTo':
+                    if (rule_spec.inputs['f'] in
+                        number_with_units_rules):
+                        raise utils.ValidationError(
+                            f'The rule \'{rule_spec_index}\' of answer '
+                            f'group \'{ans_group_index}\' has '
+                            f'rule type equal is coming after '
+                            f'rule type equivalent having same value '
+                            f'in FractionInput interaction.'
+                        )
+
+    def _validate_multi_choice_input(self) -> None:
+        """Validates the following for MultipleChoiceInput interaction -
+        - Rule should not be duplicate
+        - No ans choice should appear more than one
+        - Answer choices should be non-empty and unique
+
+        Raises:
+            ValidationError. Raises a validation error when the above
+                mentioned checks fails.
+        """
+        selected_equals_choices = []
+        rule_spec_list = []
+        rule_spec_till_now: List[state_domain.RuleSpecDict] = []
+        for ans_group_index, answer_group in enumerate(self.answer_groups):
+            for rule_spec_index, rule_spec in enumerate(
+                answer_group.rule_specs):
+                # Rule should not be duplicate.
+                if rule_spec.to_dict() in rule_spec_till_now:
+                    raise utils.ValidationError(
+                        f'The rule \'{rule_spec_index}\' of answer group '
+                        f'\'{ans_group_index}\' of MultipleChoiceInput interaction '
+                        f'is already present.'
+                    )
+                else:
+                    rule_spec_till_now.append(rule_spec.to_dict())
+                # No ans choice should appear more than one.
+                if rule_spec.rule_type == 'Equals':
+                    if rule_spec.inputs['x'] in selected_equals_choices:
+                        raise utils.ValidationError(
+                            f'The rule \'{rule_spec_index}\' of answer group '
+                            f'\'{ans_group_index}\' is already present in '
+                            f'MultipleChoiceInput interaction.'
+                        )
+                    else:
+                        selected_equals_choices.append(
+                            rule_spec.inputs['x'])
+
+                rule_spec_list.append(rule_spec.inputs['x'])
+        choices = self.customization_args['choices'].value
+
+        # Answer choices should be non-empty and unique.
+        seen_choices = []
+        choice_empty = False
+        choice_duplicate = False
+        for choice in choices:
+            if choice.html in ('<p></p>', ''):
+                choice_empty = True
+            if choice.html in seen_choices:
+                choice_duplicate = True
+            seen_choices.append(choice.html)
```

Comment:
Instead of having these boolean flags, just raise the error right away.

Reply:
Done